internal/argocd/cluster/cluster.go

```diff
@@ -39,67 +44,154 @@ func (m *Manager) UpdateClusterConnectionInfo(agentName, status string, t time.T
+// This is called when principal receives clusterCacheInfoUpdate event from agent.
+func (m *Manager) SetClusterCacheStats(clusterInfo appv1.ClusterInfo, agentName string) {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
 
-		// fetch latest info
-		clusterInfo := &appv1.ClusterInfo{}
-		if err := cache.GetClusterInfo(cluster.Server, clusterInfo); err != nil {
-			log().Errorf("Failed to get connection info from Cluster: '%s' mapped with Agent: '%s'. Error: %v", cluster.Name, agentName, err)
-			return
-		}
+	// Check if we have a mapping for the requested agent.
+	cluster := m.mapping(agentName)
+	if cluster == nil {
+		log().Errorf("agent %s is not mapped to any cluster", agentName)
+		return
+	}
+
+	// Get existing cluster info to preserve cluster connection status
+	existingClusterInfo, err := GetClusterInfo(m.ctx, m.kubeclient, m.namespace, cluster.Server, m.redisAddress, m.redisCompressionType)
+	if err != nil && !errors.Is(err, cacheutil.ErrCacheMiss) {
+		log().Errorf("failed to get existing cluster info for cluster: '%s' mapped with agent: '%s'. Error: %v", cluster.Name, agentName, err)
+		// Continue with default values if we can't get existing info
+	}
 
-		// re-save same info
-		if err := cache.SetClusterInfo(cluster.Server, clusterInfo); err != nil {
-			log().Errorf("Failed to refresh connection info in Cluster: '%s' mapped with Agent: '%s'. Error: %v", cluster.Name, agentName, err)
-			return
+	// Create new cluster info with cache stats
+	newClusterInfo := &appv1.ClusterInfo{
+		ApplicationsCount: clusterInfo.ApplicationsCount,
+		CacheInfo: appv1.ClusterCacheInfo{
+			APIsCount:      clusterInfo.CacheInfo.APIsCount,
+			ResourcesCount: clusterInfo.CacheInfo.ResourcesCount,
+		},
+	}
+
+	// Preserve existing cluster connection status
+	if existingClusterInfo != nil {
+		newClusterInfo.ConnectionState = existingClusterInfo.ConnectionState
+		if existingClusterInfo.CacheInfo.LastCacheSyncTime != nil {
+			newClusterInfo.CacheInfo.LastCacheSyncTime = existingClusterInfo.CacheInfo.LastCacheSyncTime
 		}
 	}
+
+	// Set the info in mapped cluster at principal.
+	if err := setClusterInfo(m.ctx, m.kubeclient, m.namespace, cluster.Server, m.redisAddress, agentName, cluster.Name, m.redisCompressionType, newClusterInfo); err != nil {
+		log().Errorf("failed to update cluster cache stats in cluster: '%s' mapped with agent: '%s'. Error: %v", cluster.Name, agentName, err)
+		return
+	}
+
+	log().WithFields(logrus.Fields{
+		"applicationsCount": clusterInfo.ApplicationsCount,
+		"apisCount":         clusterInfo.CacheInfo.APIsCount,
+		"resourcesCount":    clusterInfo.CacheInfo.ResourcesCount,
+		"cluster":           cluster.Name,
+		"agent":             agentName,
+	}).Infof("Updated cluster cache stats in cluster.")
+}
+
+// setClusterInfo saves the given ClusterInfo in the cache.
+func setClusterInfo(ctx context.Context, kubeclient kubernetes.Interface, namespace, clusterServer, redisAddress, agentName, clusterName string,
```

Comment:
Nit: Many of the fields that are being passed are coming from manager. Maybe we should just add this function to manager, e.g. `func (m *Manager) setClusterInfo(...)`. This would get rid of many of the parameters.